File: bc2026/src/bestsofar/Utils/BugNavigator.cpp
#include "Utils/BugNavigator.hpp"

#include <algorithm>
#include <cstdlib>

#include "Utils/FastSet.hpp"
#include "Utils/Globals.hpp"
#include "Utils/Vision.hpp"

namespace Utils {

MapLocation BugNavigator::currentTarget = MapLocation::NONE;
int BugNavigator::minDistanceToTarget = 0;
MapLocation BugNavigator::closestPositionToTarget = MapLocation::NONE;
bool BugNavigator::obstacleOnRight = false;
MapLocation BugNavigator::currentObstacle = MapLocation::NONE;
FastSet BugNavigator::visitedStates;

void BugNavigator::moveTo(MapLocation target, bool canTurn) {
    if (currentTarget.isNull() || currentTarget.distanceSquaredTo(target) > 2) {
        reset();
        closestPositionToTarget = rc.getLocation();
        minDistanceToTarget = rc.getLocation().distanceSquaredTo(target);
    } else if (!(target == currentTarget)) {
        int curDist = distance1d(rc.getLocation(), target);
        int closestDist = distance1d(closestPositionToTarget, target);
        if (curDist < closestDist) {
            minDistanceToTarget = curDist;
            closestPositionToTarget = rc.getLocation();
        } else
            minDistanceToTarget = closestDist;
    }

    if (!rc.isMovementReady()) return;

    bool hasOptions = false;
    for (int i = (int)adjacentDirections.size(); --i >= 0;) {
        if (canMoveWithDig(adjacentDirections[i])) {
            hasOptions = true;
            break;
        }
    }

    if (!hasOptions) {
        return;
    }

    MapLocation myLocation = rc.getLocation();

    int distanceToTarget = distance1d(myLocation, target);
    if (distanceToTarget < minDistanceToTarget) {
        rc.setIndicatorString("watesiggma");
        reset();
        minDistanceToTarget = distanceToTarget;
        closestPositionToTarget = rc.getLocation();
    }

    if (!currentObstacle.isNull() && Vision::hasSeenLocation(currentObstacle) &&
        Vision::sensePassability(currentObstacle)) {
        reset();
    }

    if (!visitedStates.add(getState(target))) {
        reset();
    }

    currentTarget = target;

    if (currentObstacle.isNull()) {
        Direction forward = myLocation.directionTo(target);
        if (canMoveWithDig(forward)) {
            moveWithDig(forward, canTurn);
            return;
        }

        setInitialDirection();
    }

    followWall(true, canTurn);
}

void BugNavigator::reset() {
    currentTarget = MapLocation::NONE;
    obstacleOnRight = true;
    currentObstacle = MapLocation::NONE;
    visitedStates = FastSet();
}

void BugNavigator::setInitialDirection() {
    MapLocation myLocation = rc.getLocation();
    Direction forward = myLocation.directionTo(currentTarget);

    Direction left = forward.rotateLeft();
    for (int i = 8; --i >= 0;) {
        MapLocation location = rc.adjacentLocation(left);
        if (Vision::hasSeenLocation(location) && Vision::sensePassability(location)) {
            break;
        }

        left = left.rotateLeft();
    }

    Direction right = forward.rotateRight();
    for (int i = 8; --i >= 0;) {
        MapLocation location = rc.adjacentLocation(right);
        if (Vision::hasSeenLocation(location) && Vision::sensePassability(location)) {
            break;
        }

        right = right.rotateRight();
    }

    MapLocation leftLocation = rc.adjacentLocation(left);
    MapLocation rightLocation = rc.adjacentLocation(right);

    int leftDistance = distance1d(leftLocation, currentTarget);
    int rightDistance = distance1d(rightLocation, currentTarget);

    if (leftDistance < rightDistance) {
        obstacleOnRight = true;
    } else if (rightDistance < leftDistance) {
        obstacleOnRight = false;
    } else {
        obstacleOnRight =
            myLocation.distanceSquaredTo(leftLocation) < myLocation.distanceSquaredTo(rightLocation);
    }

    if (obstacleOnRight) {
        currentObstacle = rc.adjacentLocation(left.rotateRight());
    } else {
        currentObstacle = rc.adjacentLocation(right.rotateLeft());
    }
}

void BugNavigator::followWall(bool canRotate, bool canTurn) {
    Direction direction = rc.getLocation().directionTo(currentObstacle);

    for (int i = 8; --i >= 0;) {
        direction = obstacleOnRight ? direction.rotateLeft() : direction.rotateRight();
        if (canMoveWithDig(direction)) {
            moveWithDig(direction, canTurn);
            return;
        }

        MapLocation location = rc.adjacentLocation(direction);
        if (canRotate && !rc.onTheMap(location)) {
            obstacleOnRight = !obstacleOnRight;
            followWall(false, canTurn);
            return;
        }

        if (Vision::hasSeenLocation(location) && !Vision::sensePassability(location)) {
            currentObstacle = location;
        }
    }
}

char16_t BugNavigator::getState(MapLocation target) {
    MapLocation myLocation = rc.getLocation();
    Direction direction = myLocation.directionTo(!currentObstacle.isNull() ? currentObstacle : target);
    int rotation = obstacleOnRight ? 1 : 0;

    return (char16_t)((((myLocation.x << 6) | myLocation.y) << 4) | (direction.ordinal() << 1) |
                      rotation);
}

int BugNavigator::distance1d(MapLocation a, MapLocation b) {
    return std::max(std::abs(a.x - b.x), std::abs(a.y - b.y));
}

bool BugNavigator::canMoveWithDig(Direction direction) {
    return (rc.canMove(direction) || rc.canRemoveDirt(rc.adjacentLocation(direction))) &&
           !Vision::isDangerous(rc.adjacentLocation(direction));
}

void BugNavigator::moveWithDig(Direction direction, bool canTurn) {
    MapLocation DigLocation = rc.adjacentLocation(direction);
    if (rc.canRemoveDirt(DigLocation)) {
        rc.removeDirt(DigLocation);
        Vision::digDirt(DigLocation);
    }

    if (rc.canMove(direction)) {
        if (canTurn && rc.canTurn()) {
            Vision::turn(direction);
        }
        rc.move(direction);
    }

    // Logger.log("bug " + direction);
}

}  // namespace Utils

File: py/bc2026/src/bestsofar/Utils/BugNavigator.py
# Port of bestsofar/Utils/BugNavigator.java. See py/CONVENTIONS.md.
from api import *
from Utils.FastSet import FastSet
from Utils.Globals import Globals
from Utils.Vision import Vision


class BugNavigator(Globals):
    currentTarget = None

    minDistanceToTarget = 0
    closestPositionToTarget = None

    obstacleOnRight = False
    currentObstacle = None
    visitedStates = None  # Java: private static FastSet

    @staticmethod
    def moveTo(target, canTurn):
        if BugNavigator.currentTarget is None or BugNavigator.currentTarget.distanceSquaredTo(target) > 2:
            BugNavigator.reset()
            BugNavigator.closestPositionToTarget = Globals.rc.getLocation()
            BugNavigator.minDistanceToTarget = Globals.rc.getLocation().distanceSquaredTo(target)
        elif target != BugNavigator.currentTarget:
            curDist = BugNavigator.distance1d(Globals.rc.getLocation(), target)
            closestDist = BugNavigator.distance1d(BugNavigator.closestPositionToTarget, target)
            if curDist < closestDist:
                BugNavigator.minDistanceToTarget = curDist
                BugNavigator.closestPositionToTarget = Globals.rc.getLocation()
            else:
                BugNavigator.minDistanceToTarget = closestDist

        if not Globals.rc.isMovementReady():
            return

        hasOptions = False
        # Java: for (int i = adjacentDirections.length; --i >= 0; )
        for i in range(len(Globals.adjacentDirections) - 1, -1, -1):
            if BugNavigator.canMoveWithDig(Globals.adjacentDirections[i]):
                hasOptions = True
                break

        if not hasOptions:
            return

        myLocation = Globals.rc.getLocation()

        distanceToTarget = BugNavigator.distance1d(myLocation, target)
        if distanceToTarget < BugNavigator.minDistanceToTarget:
            Globals.rc.setIndicatorString("watesiggma")
            BugNavigator.reset()
            BugNavigator.minDistanceToTarget = distanceToTarget
            BugNavigator.closestPositionToTarget = Globals.rc.getLocation()

        if BugNavigator.currentObstacle is not None and Vision.hasSeenLocation(BugNavigator.currentObstacle) and Vision.sensePassability(BugNavigator.currentObstacle):
            BugNavigator.reset()

        if not BugNavigator.visitedStates.add(BugNavigator.getState(target)):
            BugNavigator.reset()

        BugNavigator.currentTarget = target

        if BugNavigator.currentObstacle is None:
            forward = myLocation.directionTo(target)
            if BugNavigator.canMoveWithDig(forward):
                BugNavigator.moveWithDig(forward, canTurn)
                return

            BugNavigator.setInitialDirection()

        BugNavigator.followWall(True, canTurn)

    @staticmethod
    def reset():
        BugNavigator.currentTarget = None
        BugNavigator.obstacleOnRight = True
        BugNavigator.currentObstacle = None
        BugNavigator.visitedStates = FastSet()

    @staticmethod
    def setInitialDirection():
        myLocation = Globals.rc.getLocation()
        forward = myLocation.directionTo(BugNavigator.currentTarget)

        left = forward.rotateLeft()
        for i in range(7, -1, -1):  # Java: for (int i = 8; --i >= 0; )
            location = Globals.rc.adjacentLocation(left)
            if Vision.hasSeenLocation(location) and Vision.sensePassability(location):
                break

            left = left.rotateLeft()

        right = forward.rotateRight()
        for i in range(7, -1, -1):  # Java: for (int i = 8; --i >= 0; )
            location = Globals.rc.adjacentLocation(right)
            if Vision.hasSeenLocation(location) and Vision.sensePassability(location):
                break

            right = right.rotateRight()

        leftLocation = Globals.rc.adjacentLocation(left)
        rightLocation = Globals.rc.adjacentLocation(right)

        leftDistance = BugNavigator.distance1d(leftLocation, BugNavigator.currentTarget)
        rightDistance = BugNavigator.distance1d(rightLocation, BugNavigator.currentTarget)

        if leftDistance < rightDistance:
            BugNavigator.obstacleOnRight = True
        elif rightDistance < leftDistance:
            BugNavigator.obstacleOnRight = False
        else:
            BugNavigator.obstacleOnRight = myLocation.distanceSquaredTo(leftLocation) < myLocation.distanceSquaredTo(rightLocation)

        if BugNavigator.obstacleOnRight:
            BugNavigator.currentObstacle = Globals.rc.adjacentLocation(left.rotateRight())
        else:
            BugNavigator.currentObstacle = Globals.rc.adjacentLocation(right.rotateLeft())

    @staticmethod
    def followWall(canRotate, canTurn):
        direction = Globals.rc.getLocation().directionTo(BugNavigator.currentObstacle)

        for i in range(7, -1, -1):  # Java: for (int i = 8; --i >= 0; )
            direction = direction.rotateLeft() if BugNavigator.obstacleOnRight else direction.rotateRight()
            if BugNavigator.canMoveWithDig(direction):
                BugNavigator.moveWithDig(direction, canTurn)
                return

            location = Globals.rc.adjacentLocation(direction)
            if canRotate and not Globals.rc.onTheMap(location):
                BugNavigator.obstacleOnRight = not BugNavigator.obstacleOnRight
                BugNavigator.followWall(False, canTurn)
                return

            if Vision.hasSeenLocation(location) and not Vision.sensePassability(location):
                BugNavigator.currentObstacle = location

    @staticmethod
    def getState(target):
        myLocation = Globals.rc.getLocation()
        # Java: currentObstacle != null ? currentObstacle : target
        direction = myLocation.directionTo(BugNavigator.currentObstacle if BugNavigator.currentObstacle is not None else target)
        rotation = 1 if BugNavigator.obstacleOnRight else 0

        # Java: (char) cast truncates to 16 bits
        return ((((myLocation.x << 6) | myLocation.y) << 4) | (direction.ordinal() << 1) | rotation) & 0xFFFF

    @staticmethod
    def distance1d(a, b):
        return max(abs(a.x - b.x), abs(a.y - b.y))

    @staticmethod
    def canMoveWithDig(direction):
        return (Globals.rc.canMove(direction) or Globals.rc.canRemoveDirt(Globals.rc.adjacentLocation(direction))) and not Vision.isDangerous(Globals.rc.adjacentLocation(direction))

    @staticmethod
    def moveWithDig(direction, canTurn):
        DigLocation = Globals.rc.adjacentLocation(direction)
        if Globals.rc.canRemoveDirt(DigLocation):
            Globals.rc.removeDirt(DigLocation)
            Vision.digDirt(DigLocation)

        if Globals.rc.canMove(direction):
            if canTurn and Globals.rc.canTurn():
                Vision.turn(direction)
            Globals.rc.move(direction)

        # Logger.log("bug " + direction);
